File: src/js/minCameraCover.mjs
import { isEmpty } from 'lodash'

// eslint-disable-next-line @typescript-eslint/no-unused-vars
class TreeNode {
  constructor (val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}

const LEAF = 1
const CAMERA = 2
const NOCAMERA = 3

class DFS {
  constructor () {
    this.depth = 0
  }

  travel (node) {
    if (isEmpty(node)) {
      return NOCAMERA
    }

    const left = this.travel(node.left)
    const right = this.travel(node.right)
    const statuses = [left, right]

    if (statuses.includes(LEAF)) {
      this.depth += 1
      return CAMERA
    } else if (statuses.includes(CAMERA)) {
      return NOCAMERA
    }

    return LEAF
  }
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
export function minCameraCover (root) {
  const dfs = new DFS()
  const status = dfs.travel(root)
  const depth = dfs.depth
  return status === LEAF ? depth + 1 : depth
}

File: src/rs/min_camera_cover.rs
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

enum Status {
    LEAF,
    CAMERA,
    NOCAMERA,
}

struct DFS {
    depth: i32,
}

impl DFS {
    fn travel(&mut self, root: &Option<Rc<RefCell<TreeNode>>>) -> Status {
        if let Some(node) = root {
            self.travel2(node.clone())
        } else {
            Status::NOCAMERA
        }
    }

    fn travel2(&mut self, node: Rc<RefCell<TreeNode>>) -> Status {
        let left = self.travel(&node.borrow().left);
        let right = self.travel(&node.borrow().right);
        match (left, right) {
            (Status::LEAF, _) | (_, Status::LEAF) => {
                self.depth += 1;
                Status::CAMERA
            }
            (Status::CAMERA, _) | (_, Status::CAMERA) => Status::NOCAMERA,
            _ => Status::LEAF,
        }
    }
}

pub struct MinCameraCover {}
impl MinCameraCover {
    pub fn min_camera_cover(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut dfs = DFS { depth: 0 };
        let status = dfs.travel(&root);
        let depth = dfs.depth;
        match status {
            Status::LEAF => depth + 1,
            _ => depth,
        }
    }
}
